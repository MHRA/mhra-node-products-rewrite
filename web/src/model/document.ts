export interface IDocument {
  activeSubstances: string[];
  context: string;
  created: string;
  docType: string;
  fileSize: string;
  name: string;
  product: string;
  url: string;
}

export interface IDocuments {
  count: number;
  documents: IDocument[];
}

export interface IBmgfReport {
  activeSubstances: string[];
  context: string;
  fileName: string;
  fileSize: string;
  fileUrl: string;
  title: string;
  products: string[];
  matrices: string[];
  pbpkModels: string[];
  summary: string;
  url: string;
  pregnancyTrimesters: string[];
  plNumbers: string[];
}

export interface IBmgfReports {
  count: number;
  reports: IBmgfReport[];
}
