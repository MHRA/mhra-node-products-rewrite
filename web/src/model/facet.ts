export interface IFacet {
  name: string;
  count?: number;
}
