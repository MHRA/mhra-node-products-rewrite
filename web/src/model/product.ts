export interface IProduct {
  name: string;
  count?: number;
}
