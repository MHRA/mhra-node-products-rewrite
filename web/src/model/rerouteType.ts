export enum RerouteType {
  CheckboxSelected,
  Other,
}
