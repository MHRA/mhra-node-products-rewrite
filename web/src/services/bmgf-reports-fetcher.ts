import {
  BlobServiceClient,
  StorageSharedKeyCredential,
} from '@azure/storage-blob';

const containerUrl = `https://${process.env.AZURE_STORAGE_ACCOUNT}.blob.core.windows.net/bmgf-docs`;
const account = process.env.AZURE_STORAGE_ACCOUNT || '';
const accountKey = process.env.AZURE_STORAGE_KEY || '';

export const getReportList = async (): Promise<string[]> => {
  const sharedKeyCredential = new StorageSharedKeyCredential(
    account,
    accountKey,
  );
  const blobServiceClient = new BlobServiceClient(
    `https://${account}.blob.core.windows.net`,
    sharedKeyCredential,
  );

  const containerName = 'bmgf-docs';
  const containerClient = blobServiceClient.getContainerClient(containerName);

  const blobs = containerClient.listBlobsFlat();
  const reportNames = [];
  for await (const blob of blobs) {
    if (blob.name.endsWith('.html')) {
      reportNames.push(blob.name);
    }
  }
  return reportNames;
};

export const getReportUrls = async (
  reportName: string,
  fileNames: string[],
): Promise<any> => {
  const fileName = fileNames.find((fileName) =>
    fileName.startsWith(reportName),
  );
  return {
    reportPdfUrl: `${containerUrl}/${fileName?.replace('.html', '.pdf')}`,
    reportHtmlUrl: `${containerUrl}/${fileName}`,
    reportAssetsUrl: `${containerUrl}/${reportName}/assets/`,
  };
};

export const getReportHtmlContent = async (
  reportUrl: string,
): Promise<string> => {
  return fetch(`${reportUrl}`, {
    method: 'GET',
    headers: {
      'Content-Type': 'text/html',
    },
  }).then((response) => (response.ok ? response.text() : ''));
};
