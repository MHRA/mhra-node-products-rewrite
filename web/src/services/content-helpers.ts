export const pluralise = (
  singular: string,
  plural: string,
  count: number,
): string => {
  return count === 1 ? singular : plural;
};
