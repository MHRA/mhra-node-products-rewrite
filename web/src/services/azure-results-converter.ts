import moment from 'moment';

import { IDocument, IBmgfReport } from '../model/document';
import { IFacet } from '../model/facet';
import {
  IFacetResult,
  ISearchResult,
  IBmgfSearchResult,
} from '../services/azure-search';

const sanitizeTitle = (title: string | null): string => {
  let name: string;
  if (!title) return 'Unknown';

  try {
    name = decodeURIComponent(title);
  } catch {
    name = title;
  }
  return name;
};

export const convertResults = (doc: ISearchResult): IDocument => {
  return {
    activeSubstances: doc.substance_name,
    product: doc.product_name,
    context: doc['@search.highlights']?.content.join(' … ') || '',
    docType: doc.doc_type?.toString().substr(0, 3) || '',
    fileSize: Math.ceil(
      (doc.metadata_storage_size ? doc.metadata_storage_size : 0) / 1000,
    ).toLocaleString('en-GB'),
    created: doc.created
      ? moment(doc.created).format('D MMMM YYYY')
      : 'Unknown',
    name: sanitizeTitle(doc.title),
    url: doc.metadata_storage_path,
  };
};

export const convertBmgfResults = (doc: IBmgfSearchResult): IBmgfReport => {
  return {
    activeSubstances: doc.active_substances,
    context: doc['@search.highlights']?.content.join(' … ') || '',
    fileName: doc.file_name || '',
    fileUrl: doc.metadata_storage_path,
    products: doc.products,
    summary: doc.summary,
    pbpkModels: doc.pbpk_models,
    matrices: doc.matrices,
    title: sanitizeTitle(doc.report_name),
    fileSize: Math.ceil(
      (doc.metadata_storage_size ? doc.metadata_storage_size : 0) / 1000,
    ).toLocaleString('en-GB'),
    url: `/medicine-levels-in-pregnancy/reports/${doc.report_name}`,
    plNumbers: doc.pl_numbers,
    pregnancyTrimesters: doc.pregnancy_trimesters,
  };
};

export const mapSubstancesIndex = ([letter, facetResult]: [
  string,
  IFacetResult,
]): IFacet[] => {
  const indexResults: { [id: string]: IFacet } = {};
  facetResult.facets
    .filter((x) => x.value.startsWith(letter))
    .forEach((f) => {
      const [substance, product] = f.value
        .replace(/\s+/g, ' ')
        .split(', ', 3)
        .slice(1);

      if (substance && !product) {
        const currentSubstanceCount = indexResults[substance]?.count ?? 0;
        indexResults[substance] = {
          name: substance,
          count: currentSubstanceCount + f.count,
        };
      }
    });
  return Object.values(indexResults).sort((a, b) => (a.name < b.name ? -1 : 1));
};

export const mapProductsIndex = ([letterAndSubstance, facetResult]: [
  string,
  IFacetResult,
]): IFacet[] => {
  const indexResults: { [id: string]: IFacet } = {};
  const requiredFacetPrefix = `${letterAndSubstance},`;
  facetResult.facets.forEach((f) => {
    const sanitizedFacet = f.value.replace(/\s+/g, ' ');
    if (!sanitizedFacet.startsWith(requiredFacetPrefix)) {
      return;
    }

    const [product] = sanitizedFacet.split(', ', 3).slice(2);

    if (product) {
      const currentProductCount = indexResults[product]?.count ?? 0;
      indexResults[product] = {
        name: product,
        count: currentProductCount + f.count,
      };
    }
  });
  return Object.values(indexResults).sort((a, b) => (a.name < b.name ? -1 : 1));
};
