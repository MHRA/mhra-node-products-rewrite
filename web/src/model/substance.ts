import { IProduct } from './product';

export interface ISubstance extends IProduct {
  products?: IProduct[];
}

export function isSubstance(obj: any): obj is ISubstance {
  return obj.products !== undefined && obj.count !== undefined;
}

export function isIndex(obj: any): obj is ISubstance {
  return obj.products === undefined && obj.count === undefined;
}
