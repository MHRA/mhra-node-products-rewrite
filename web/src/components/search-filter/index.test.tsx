import renderer from 'react-test-renderer';
import React from 'react';
import { RerouteType } from '../../model/rerouteType';
import { DocType, TerritoryType } from '../../services/azure-search';
import SearchFilter from './index';

describe(SearchFilter, () => {
  it('should render search filter', () => {
    const updatePageFiltersFunction = (d: DocType[], t: TerritoryType[]) => {
      return null;
    };
    const component = renderer
      .create(
        <SearchFilter
          currentlyEnabledDocTypes={[]}
          currentlyEnabledTerritoryTypes={[]}
          updatePageFilters={updatePageFiltersFunction}
          rerouteType={RerouteType.CheckboxSelected}
        />,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
