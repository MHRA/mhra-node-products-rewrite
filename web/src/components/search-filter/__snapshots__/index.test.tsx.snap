// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchFilter should render search filter 1`] = `
<section
  className="sc-dkzDqf gFDbOp"
>
  <h3
    className="sc-iBkjds jopVIF"
  >
    Documents filter
  </h3>
  <fieldset
    className="sc-eCYdqJ kcQGTi"
  >
    <legend
      className="sc-jSMfEi hgJpQT"
    >
      Filter documents by
    </legend>
    <legend
      className="sc-gKXOVf gQrprt"
    >
      Type of document
    </legend>
    <div
      className="checkbox-row"
    >
      <div
        className="checkbox"
      >
        <input
          checked={false}
          className="sc-bczRLJ ldzjBD"
          id="filter-spc"
          name="doc"
          onChange={[Function]}
          type="checkbox"
          value="Spc"
        />
      </div>
      <label
        htmlFor="filter-spc"
      >
        Summary of Product Characteristics
         (
        SPC
        )
      </label>
    </div>
    <div
      className="checkbox-row"
    >
      <div
        className="checkbox"
      >
        <input
          checked={false}
          className="sc-bczRLJ ldzjBD"
          id="filter-pil"
          name="doc"
          onChange={[Function]}
          type="checkbox"
          value="Pil"
        />
      </div>
      <label
        htmlFor="filter-pil"
      >
        Patient Information Leaflet
         (
        PIL
        )
      </label>
    </div>
    <div
      className="checkbox-row"
    >
      <div
        className="checkbox"
      >
        <input
          checked={false}
          className="sc-bczRLJ ldzjBD"
          id="filter-par"
          name="doc"
          onChange={[Function]}
          type="checkbox"
          value="Par"
        />
      </div>
      <label
        htmlFor="filter-par"
      >
        Public Assessment Reports
         (
        PAR
        )
      </label>
    </div>
    <legend
      className="sc-gKXOVf gQrprt"
    >
      Applicable to territory
    </legend>
    <div
      className="checkbox-row"
    >
      <div
        className="checkbox"
      >
        <input
          checked={false}
          className="sc-bczRLJ ldzjBD"
          id="filter-uk"
          name="ter"
          onChange={[Function]}
          type="checkbox"
          value="UK"
        />
      </div>
      <label
        htmlFor="filter-uk"
      >
        United Kingdom
         (
        UK
        )
      </label>
    </div>
    <div
      className="checkbox-row"
    >
      <div
        className="checkbox"
      >
        <input
          checked={false}
          className="sc-bczRLJ ldzjBD"
          id="filter-ni"
          name="ter"
          onChange={[Function]}
          type="checkbox"
          value="NI"
        />
      </div>
      <label
        htmlFor="filter-ni"
      >
        Northern Ireland
         (
        NI
        )
      </label>
    </div>
    <div
      className="checkbox-row"
    >
      <div
        className="checkbox"
      >
        <input
          checked={false}
          className="sc-bczRLJ ldzjBD"
          id="filter-gb"
          name="ter"
          onChange={[Function]}
          type="checkbox"
          value="GB"
        />
      </div>
      <label
        htmlFor="filter-gb"
      >
        Great Britain
         (
        GB
        )
      </label>
    </div>
    <div
      className="sc-hKMtZM iefjqC"
    >
      <input
        className="sc-gsnTZi jmRmnt"
        onClick={[Function]}
        type="submit"
        value="Submit"
      />
    </div>
  </fieldset>
</section>
`;
